lib/instrumentation/async-hooks.js

```diff
@@ -3,9 +3,19 @@
 const asyncHooks = require('async_hooks')
 
 module.exports = function (ins) {
-  const asyncHook = asyncHooks.createHook({init, before, after, destroy})
-  const initState = new Map()
-  const beforeState = new Map()
+  const asyncHook = asyncHooks.createHook({init, destroy})
+  const transactions = new Map()
+
+  Object.defineProperty(ins, 'currentTransaction', {
+    get: function () {
+      const asyncId = asyncHooks.executionAsyncId()
+      return transactions.get(asyncId)
```

Comment:
Not sure if it's still the case, but previously `map.get(...)` could deopt if the item did not exist. It might be a good idea to put a `if (!transactions.has(asyncId)) return` before the get.